src/main.rs

```diff
@@ -194,15 +216,40 @@ fn create_swarm(identity: identity::Keypair, ping: bool) -> Result<Swarm<Behavio
     Ok(swarm)
 }
 
-fn create_transport(identity: &identity::Keypair) -> Result<Boxed<(PeerId, StreamMuxerBox)>> {
+fn create_transport(
+    identity: &identity::Keypair,
+    websocket: bool,
+) -> Result<Boxed<(PeerId, StreamMuxerBox)>> {
+    let tcp_with_dns = TokioDnsConfig::system(TokioTcpConfig::new().nodelay(true)).unwrap();
+
+    let transport = if websocket {
+        let websocket_with_dns = WsConfig::new(tcp_with_dns.clone());
+        authenticate_and_multiplex(
+            tcp_with_dns.or_transport(websocket_with_dns).boxed(),
+            &identity,
+        )
+        .unwrap()
+    } else {
+        authenticate_and_multiplex(tcp_with_dns.boxed(), &identity).unwrap()
+    };
+
+    Ok(transport)
+}
+
+fn authenticate_and_multiplex<T>(
+    transport: Boxed<T>,
+    identity: &identity::Keypair,
+) -> Result<Boxed<(PeerId, StreamMuxerBox)>>
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
     let auth_upgrade = {
         let noise_identity = noise::Keypair::<X25519Spec>::new().into_authentic(identity)?;
         NoiseConfig::xx(noise_identity).into_authenticated()
     };
     let multiplex_upgrade = SelectUpgrade::new(YamuxConfig::default(), MplexConfig::new());
 
-    let transport = TokioDnsConfig::system(TokioTcpConfig::new().nodelay(true))
-        .context("Failed to create DNS transport")?
+    let transport = transport
```

Comment:
Using [`OptionalTransport`](https://docs.rs/libp2p/0.39.1/libp2p/core/transport/struct.OptionalTransport.html) would likely be a less invasive change, result in less code and be slightly more performant because you don't need to call `.boxed()` on the transport.

Reply:
happy to see this in a follow up